### PR TITLE
Unify and handle exceptions

### PR DESCRIPTION
**Pull request by jonasfj**
This is highly inspired by the PR @isoos filed recently.

### CODE REVIEW

**Comment by jonasfj** [comment]
@isoos, there is other exception types defined in `pub_server` we should make subclasses of those, and move them all into `exceptions.dart`. But let's do that in a later PR..

File: app/lib/shared/exceptions.dart [hunk 1]
```diff
@@ -2,21 +2,223 @@
...
+  /// than [maximum].
+  static void checkRange<T extends num>(
+    T value,
+    String name, {
+    T minimum,
+    T maximum,
+  }) {
+    _check(value != null, () => '"$name" cannot be `null`');
+    assert(name != null, '"name" must not be `null`');
+    _check((minimum == null || value >= minimum),
```
Comment (isoos) [inline app/lib/shared/exceptions.dart]: It is a bit confusing to read this, I was excepting to throw the exception when the condition is true. Maybe `_assertOrThrow`?

File: app/lib/shared/exceptions.dart [hunk 2]
```diff
@@ -2,21 +2,223 @@
...
+      );
 
-/// Exception when a package or a version is missing, or has other flags that
-/// indicate it should be removed from the search index.
-class RemovedPackageException extends NotFoundException {}
+  /// Signaling that the user does not have permissions to upload a new version
+  /// of [package].
+  factory AuthorizationException.userCannotUploadNewVersion(String package) =>
+      AuthorizationException._(
+        'Insufficient permissions to upload new versions of package `$package`',
```
Comment (isoos) [inline app/lib/shared/exceptions.dart]: Nit: ending .

File: app/lib/shared/exceptions.dart [hunk 3]
```diff
@@ -2,21 +2,223 @@
...
+      );
+
+  /// Signaling that the user does not have permissions to change uploaders for
+  /// given [package].
+  factory AuthorizationException.userCannotChangeUploaders(String package) =>
+      AuthorizationException._(
+        'Unsufficient permissions to change uploaders for `$package`',
```
Comment (isoos) [inline app/lib/shared/exceptions.dart]: Nit: ending .

File: app/lib/shared/exceptions.dart [hunk 4]
```diff
@@ -2,21 +2,223 @@
...
+  /// Create a [ConflictException] with a [message] explaining what the conflict
+  /// is and how to resolve it.
+  ConflictException._(String message)
+      : super._(409, 'RequestConflict', message);
+}
+
+/// Thrown when the analysis for a package is not done yet.
+class MissingAnalysisException extends NotFoundException {
+  MissingAnalysisException()
+      : super('Analysis is not ready for the given package');
```
Comment (isoos) [inline app/lib/shared/exceptions.dart]: Nit: ending .

File: app/lib/shared/exceptions.dart [hunk 5]
```diff
@@ -2,21 +2,223 @@
...
+}
+
+/// Thrown when package or versions is missing or has flags indicating that it
+/// should be removed from the search index.
+class RemovedPackageException extends NotFoundException {
+  RemovedPackageException() : super('Package has been removed');
```
Comment (isoos) [inline app/lib/shared/exceptions.dart]: Nit: ending .

File: app/lib/shared/exceptions.dart (hunk 1 above)
Comment (jonasfj) [reply app/lib/shared/exceptions.dart]: Checks that the condition is true... maybe we could name it `_ensure` but it's really named after the public variant..

**Comment by isoos** [comment]
Landing this with the failing publisher backend tests, as they are (a) not in use yet, and (b) will be refactored in #2514 (and there is already enough merge conflict between the two, not worth to fix the tests.)

**Pull request merged**